streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -648,55 +648,61 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
-                                  )
-                                }.foldCauseZIO(
-                                  _ => p.refailCause(Cause.unit) *> errorSignal.refailCause(Cause.unit),
-                                  p.succeed
+                   cause.failureOrCause match {
+                     case Left(Left(outErr)) =>
+                       failure.update(_ && Cause.fail(outErr)) *>
+                         outgoing.offer(Fiber.fail(Left(()))) *>
+                         Exit.failCause(Cause.unit)
+                     case Left(Right(outDone)) =>
+                       permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(
+                         Fiber.fail(Right(outDone))
+                       )
+                     case Right(cause) =>
+                       failure.update(_ && cause) *>
+                         outgoing.offer(Fiber.fail(Left(()))) *>
+                         Exit.failCause(Cause.unit)
+                   },
+                 outElem =>
+                   for {
+                     latch <- Promise.make[Nothing, Unit]
+                     f <- permits.withPermit {
+                            latch.succeed(()) *>
+                              ZIO.uninterruptibleMask { restore =>
+                                restore(errorSignal.await).raceFirstAwait(
+                                  restore(f(outElem))
+                                    .catchAllCause(cause => failure.update(_ && cause) *> Exit.failCause(Cause.unit))
                                 )
-                            }.forkIn(scope)
-                       _ <- latch.await
-                     } yield ()
-                 }
+                              }.tapErrorCause(_ => errorSignal.refailCause(Cause.unit))
+                                .mapError(Left(_))
+                          }.forkIn(scope)
+                     _ <- outgoing.offer(f)
+                     _ <- latch.await
+                   } yield ()
                )
                .forever
                .interruptible
                .forkIn(scope)
       } yield {
         lazy val writer: ZChannel[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] =
           ZChannel.unwrap[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] {
-            outgoing.take.flatten.foldCause(
-              _ => ZChannel.unwrap(failure.get.map(ZChannel.refailCause(_))),
-              {
-                case Left(outDone)  => ZChannel.succeedNow(outDone)
-                case Right(outElem) => ZChannel.write(outElem) *> writer
-              }
-            )
+            outgoing.take.flatMap(_.await).map {
+              case s: Exit.Success[OutElem2] => ZChannel.write(s.value) *> writer
+              case f: Exit.Failure[Either[Unit, OutDone]] =>
+                f.cause.failureOrCause match {
+                  case Left(Left(()))       => ZChannel.unwrap(failure.get.map(ZChannel.refailCause(_)))
```

Comment:
I think you want to avoid matching on the Unit to avoid another equality check.

Reply:
Good idea, done